### DISCUSSION TITLE
DEPLOY FIEF ON GCP

### DISCUSSION THREAD

**Discussion by MathieuGrosso** [fief-dev/fief#336] [Q&A] [answered]
```
[supervisord]
nodaemon=true
logfile=/tmp/supervisord.log
pidfile=/tmp/supervisord.pid
childlogdir=/tmp

[program:fief-server]
command=/bin/zsh -c "fief run-server --port 8000"
autostart=true
autorestart=true
stdout_logfile=/dev/stdout
stdout_logfile_maxbytes=0
stderr_logfile=/dev/stderr
stderr_logfile_maxbytes=0

[program:fief-worker]
command=/bin/zsh -c "fief run-worker -p 1 -t 1"
autostart=true
autorestart=true
stdout_logfile=/dev/stdout
stdout_logfile_maxbytes=0
stderr_logfile=/dev/stderr
stderr_logfile_maxbytes=0

[program:redis]
command=/usr/local/bin/redis-server --save 60 1 --loglevel warning
autostart=true
autorestart=true
stdout_logfile=/dev/stdout
stdout_logfile_maxbytes=0
stderr_logfile=/dev/stderr
stderr_logfile_maxbytes=0
```

hello i am trying to deploy fief directly on gcp vm since my db is on a VPC. I need to deploy fief in the vpc so in one of our gcp vm. 
I am struggling to deploy redis on the vm with supervisord. Indeed, compute engine does not accept docker compose so i need to use supervisord to make it work 
If by any chance someone knows how to do this let me know. 

---- French version !

Bonjour ! 
On essaie de déployer fief, on voulait essayer ce package pour gérer l'auth mais on a du mal à le déployer sur gcp avec supervisord. Le principal suspect étant redis. Pour l'instant voilà l'erreur: 

```
> 2024-02-14 16:48:11,887 INFO supervisord started with pid 6087
> 2024-02-14 16:48:12,892 INFO spawned: 'fief-server' with pid 6092
> 2024-02-14 16:48:12,895 INFO spawned: 'fief-worker' with pid 6093
> 2024-02-14 16:48:13,897 INFO success: fief-server entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
> 2024-02-14 16:48:13,897 INFO success: fief-worker entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
> INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
> INFO  [alembic.runtime.migration] Will assume transactional DDL.
> [2024-02-14 16:48:19,694] [PID 6093] [MainThread] [dramatiq.MainProcess] [INFO] Dramatiq '1.14.2' is booting up.
> 2024-02-14 16:48:19.452 | INFO     | fief.worker:<module>:19 - Fief Worker started - {"version": "0.27.0"}
> 2024-02-14 16:48:19.684 | INFO     | dramatiq.cli:worker_process:412 - Worker process is ready for action. - {}
> 2024-02-14 16:48:19.736 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 8 connecting to redis:6379. nodename nor servname provided, or not known. - {}
> 2024-02-14 16:48:19.737 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
> 2024-02-14 16:48:19.738 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 8 connecting to redis:6379. nodename nor servname provided, or not known. - {}
> 2024-02-14 16:48:19.739 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
> [2024-02-14 16:48:20,214] [PID 6105] [MainThread] [dramatiq.ForkProcess(1)] [INFO] Fork process 'dramatiq.middleware.prometheus:_run_exposition_server' is ready for action.
> Main Fief workspace already exists
> [2024-02-14 16:48:21,160] [PID 6104] [MainThread] [dramatiq.ForkProcess(0)] [INFO] Fork process 'fief.scheduler:schedule' is ready for action.
> [2024-02-14 16:48:21,166] [PID 6104] [MainThread] [apscheduler.scheduler] [INFO] Adding job tentatively -- it will be properly scheduled when the scheduler starts
> [2024-02-14 16:48:21,166] [PID 6104] [MainThread] [apscheduler.scheduler] [INFO] Adding job tentatively -- it will be properly scheduled when the scheduler starts
> [2024-02-14 16:48:21,166] [PID 6104] [MainThread] [apscheduler.scheduler] [INFO] Added job "Actor.send" to job store "default"
> [2024-02-14 16:48:21,166] [PID 6104] [MainThread] [apscheduler.scheduler] [INFO] Added job "Actor.send" to job store "default"
> [2024-02-14 16:48:21,166] [PID 6104] [MainThread] [apscheduler.scheduler] [INFO] Scheduler started
> Main Fief user already exists
> Main Fief admin API key not provided in settings. Skipping its creation.
> 2024-02-14 16:48:22.740 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 8 connecting to redis:6379. nodename nor servname provided, or not known. - {}
> 2024-02-14 16:48:22.741 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
> 2024-02-14 16:48:22.743 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 8 connecting to redis:6379. nodename nor servname provided, or not known. - {}
> 2024-02-14 16:48:22.743 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
> INFO:     Started server process [6092]
> INFO:     Waiting for application startup.
> 2024-02-14 16:48:24.443 | INFO     | fief.lifespan:lifespan:28 - Fief Server started - {"version": "0.27.0"}
> 2024-02-14 16:48:24.443 | WARNING  | fief.lifespan:lifespan:31 - Telemetry is enabled.
> We will collect data to better understand how Fief is used and improve the project.
> You can opt-out by setting the environment variable `TELEMETRY_ENABLED=false`.
> Read more about Fief's telemetry here: https://docs.fief.dev/telemetry - {}
> 2024-02-14 16:48:25.751 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 8 connecting to redis:6379. nodename nor servname provided, or not known. - {}
> 2024-02-14 16:48:25.751 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
> 2024-02-14 16:48:25.755 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 8 connecting to redis:6379. nodename nor servname provided, or not known. - {}
> 2024-02-14 16:48:25.756 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
> 2024-02-14 16:48:28.759 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 8 connecting to redis:6379. nodename nor servname provided, or not known. - {}
> 2024-02-14 16:48:28.761 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
> 2024-02-14 16:48:28.765 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 8 connecting to redis:6379. nodename nor servname provided, or not known. - {}
> 2024-02-14 16:48:28.766 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
> 2024-02-14 16:48:31.769 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 8 connecting to redis:6379. nodename nor servname provided, or not known. - {}
> 2024-02-14 16:48:31.770 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
> 2024-02-14 16:48:31.772 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 8 connecting to redis:6379. nodename nor servname provided, or not known. - {}
> 2024-02-14 16:48:31.773 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
> 2024-02-14 16:48:34.779 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 8 connecting to redis:6379. nodename nor servname provided, or not known. - {}
```

Si quelqu'un a une idée, faites le nous savoir ! 

Bonne journée.

**Comment by fief-bailiff** [comment]
Hail, @MathieuGrosso 👋 Welcome to Fief's kingdom!

Our team will get back to you very soon to help.

In the meantime, take a minute to star our repository ⭐️

![star-fief](https://user-images.githubusercontent.com/1144727/192523552-dc2c31fa-ae94-42fa-9ab4-7d0ba51bb7b0.png)

Farewell!

**Comment by frankie567** [comment]
Salut @MathieuGrosso 👋 

If I understand correctly, you're running Redis on the same machine/VM than Fief itself, processes being managed by supervisord. From the logs, it seems that Fief tries to reach Redis using the hostname `redis`:

```
Error 8 connecting to redis:6379. nodename nor servname provided, or not known. - {}
```

If Redis does indeed run on the same machine/VM, then you probably should reach it directly on `localhost`. So, I suggest you to try the following environment variable:

```
REDIS_URL=redis://localhost:6379
```

*(which is actually the default if you don't set it)*

**Comment by MathieuGrosso** [comment]
Salut François 🙋🙋🙋!

I just tried what you said, it seems to work locally but not on a compute engine vm. 

I get this error: 

```
2024-02-14 17:04:18,736 CRIT Supervisor is running as root.  Privileges were not dropped because no user is specified in the config file.  If you intend to run as root, you can set user=root in the config file to avoid this message.
2024-02-14 17:04:18,737 INFO supervisord started with pid 1
2024-02-14 17:04:19,739 INFO spawned: 'fief-server' with pid 7
2024-02-14 17:04:19,742 INFO spawned: 'fief-worker' with pid 8
2024-02-14 17:04:19,742 INFO spawnerr: can't find command '/usr/local/bin/redis-server'
2024-02-14 17:04:20,743 INFO success: fief-server entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-02-14 17:04:20,744 INFO success: fief-worker entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-02-14 17:04:20,744 INFO spawnerr: can't find command '/usr/local/bin/redis-server'
2024-02-14 17:04:22,745 INFO spawnerr: can't find command '/usr/local/bin/redis-server'
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
2024-02-14 17:04:25,894 INFO spawnerr: can't find command '/usr/local/bin/redis-server'
2024-02-14 17:04:25,895 INFO gave up: redis entered FATAL state, too many start retries too quickly
[2024-02-14 17:04:26,397] [PID 8] [MainThread] [dramatiq.MainProcess] [INFO] Dramatiq '1.14.2' is booting up.
2024-02-14 17:04:26.259 | INFO     | fief.worker:<module>:19 - Fief Worker started - {"version": "0.27.0"}
2024-02-14 17:04:26.390 | INFO     | dramatiq.cli:worker_process:412 - Worker process is ready for action. - {}
[2024-02-14 17:04:26,400] [PID 16] [MainThread] [dramatiq.ForkProcess(1)] [INFO] Fork process 'dramatiq.middleware.prometheus:_run_exposition_server' is ready for action.
2024-02-14 17:04:26.409 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 111 connecting to localhost:6379. Connection refused. - {}
2024-02-14 17:04:26.409 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
2024-02-14 17:04:26.410 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 111 connecting to localhost:6379. Connection refused. - {}
2024-02-14 17:04:26.410 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
Main Fief workspace already exists
Main Fief user already exists
Main Fief admin API key not provided in settings. Skipping its creation.
[2024-02-14 17:04:27,974] [PID 15] [MainThread] [dramatiq.ForkProcess(0)] [INFO] Fork process 'fief.scheduler:schedule' is ready for action.
[2024-02-14 17:04:27,978] [PID 15] [MainThread] [apscheduler.scheduler] [INFO] Adding job tentatively -- it will be properly scheduled when the scheduler starts
[2024-02-14 17:04:27,978] [PID 15] [MainThread] [apscheduler.scheduler] [INFO] Adding job tentatively -- it will be properly scheduled when the scheduler starts
[2024-02-14 17:04:27,979] [PID 15] [MainThread] [apscheduler.scheduler] [INFO] Added job "Actor.send" to job store "default"
[2024-02-14 17:04:27,980] [PID 15] [MainThread] [apscheduler.scheduler] [INFO] Added job "Actor.send" to job store "default"
[2024-02-14 17:04:27,980] [PID 15] [MainThread] [apscheduler.scheduler] [INFO] Scheduler started
2024-02-14 17:04:29.411 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 111 connecting to localhost:6379. Connection refused. - {}
2024-02-14 17:04:29.412 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
2024-02-14 17:04:29.413 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 111 connecting to localhost:6379. Connection refused. - {}
2024-02-14 17:04:29.413 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
INFO:     Started server process [7]
INFO:     Waiting for application startup.
2024-02-14 17:04:32.413 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 111 connecting to localhost:6379. Connection refused. - {}
2024-02-14 17:04:32.413 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
2024-02-14 17:04:32.414 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 111 connecting to localhost:6379. Connection refused. - {}
2024-02-14 17:04:32.414 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
2024-02-14 17:04:32.569 | INFO     | fief.lifespan:lifespan:28 - Fief Server started - {"version": "0.27.0"}
2024-02-14 17:04:32.569 | WARNING  | fief.lifespan:lifespan:31 - Telemetry is enabled.
We will collect data to better understand how Fief is used and improve the project.
You can opt-out by setting the environment variable `TELEMETRY_ENABLED=false`.
Read more about Fief's telemetry here: https://docs.fief.dev/telemetry - {}
2024-02-14 17:04:35.414 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 111 connecting to localhost:6379. Connection refused. - {}
2024-02-14 17:04:35.414 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
2024-02-14 17:04:35.415 | CRITICAL | dramatiq.worker:run:276 - Consumer encountered a connection error: Error 111 connecting to localhost:6379. Connection refused. - {}
2024-02-14 17:04:35.415 | INFO     | dramatiq.worker:run:288 - Restarting consumer in 3.00 seconds. - {}
```

**Reply by MathieuGrosso** [reply]
It feels like the vm cannot find the redis instane while it has been installed. 
Maybe this is something more linked to compute engine way of working  and not fief directly

**Reply by MathieuGrosso** [reply]
Just to precise: I am using a dockerfile to launch the supervisord conf in the vm 

```
# use a ubuntu image and avoid the interactive installation
FROM python:3.11.4 AS base
# Setup env vars to put python and pip into prod mode in order to not generate any
WORKDIR /###/libs/###


# Set environment variables
ENV GCP_PROJECT="###"
ENV GOOGLE_CLOUD_PROJECT="##"

ENV PORT=8000 \
    REDIS_PORT=6379 \
    GCP_PROJECT="###" \
    GOOGLE_CLOUD_PROJECT="###"

# Install Python and pip
RUN apt update && apt install -y \
    python3-pip \
    supervisor postgresql \
    nano zsh screen neovim bzip2 libx11-6 vim unzip build-essential ssh && \
    rm -rf /tmp/* && \
    rm -rf /var/lib/apt/lists/* &&\
    mkdir -p ~/.ssh && ssh-keyscan github.com >> ~/.ssh/known_hosts

RUN pip3 install psycopg2 fief-server fief-client

################################################################
#                    supervisord and redis                     #
################################################################
RUN apt-get update && apt-get install -y supervisor redis-server && rm -rf /var/lib/apt/lists/* \
    && redis-server --version


COPY . /##
COPY ./libs/##/.env .env 
COPY ./libs/##/supervisord.conf /etc/supervisor/conf.d/supervisord.conf

# wee need to expose th prt
EXPOSE ${PORT}

# Start Supervisor when the container starts
CMD ["/usr/bin/supervisord", "-c", "/etc/supervisor/conf.d/supervisord.conf"]

```

**Reply by frankie567** [reply]
Indeed, for some reasons, the system can't start Redis:

```
2024-02-14 17:04:19,742 INFO spawnerr: can't find command '/usr/local/bin/redis-server'
```

Not sure if that could help but, in our official image, we install the `redis` package, not `redis-server`:

https://github.com/fief-dev/fief/blob/a08dcd7253c3ece8781ace7512324561848b9723/docker/Dockerfile#L8

**Reply by frankie567** [reply]
Side note: if you don't want the hassle of managing a Redis server yourself, you can check out Redis Cloud: https://redis.com/cloud/overview/

They create and manage an instance on the cloud of your choice (like GCP). They seem to support VPC: https://docs.redis.com/latest/rc/security/vpc-peering/ Mind the costs though, it's probably not cheap 😬

**Reply by MathieuGrosso** [reply]
Hello ! I have been able to get it working :) 
a malformed redis was the issue - indeed an old version of redis was staying in the volume :) 

I still have an issue with the domain, and i am a bit surprise by it. 
I have put localhost:8000 as the config for fief domain and root domain. When i ssh to the vm, i see the fief login page on localhost:8000. 


```
PORT=8000
ROOT_DOMAIN=localhost:8000
FIEF_DOMAIN=localhost:8000

```

But then i wanted to assign it to a domain, so i created a domain name for my vm. I was surprised to see that now the external ip of the vm port 8000, i get : {"detail":"CANT_DETERMINE_VALID_WORKSPACE"}. 
(address: http://35.195.72.191:8000/).



I dont know if i need to update the config + i am not sure to understand what are those values: 

```

PORT=8000
ROOT_DOMAIN=35.195.72.191
FIEF_DOMAIN=35.195.72.191

```

I have tried like this, but then i cant access anything... 


THanks for the help !

**Reply by frankie567** [reply]
Great!

Well, domain can indeed be a bit complex to get up and running (that's why I'm currently [revamping this right now](https://github.com/orgs/fief-dev/discussions/322)).

The right value for `ROOT_DOMAIN ` and `FIEF_DOMAIN` should be: `35.195.72.191:8000`. It must match the full hostname used to access the instance.

Be careful though: when you change those values, it'll create a new workspace automatically at startup. That's because, in the current version, we have a "map" of workspaces and their domains **inside the database** (because we support multiple workspaces — that's what I'll remove soon).

So, if you change the domain but want to keep using the existing workspace, you'll need to look for the `workspaces` table inside your DB and change the `domain` column.

---

I suppose that, in the end, you'll want to serve it from a proper subdomain, **with a proper HTTPS reverse proxy (that's critical)**. Let's say the domain will be `fief.example.com`. Then `ROOT_DOMAIN`, `FIEF_DOMAIN` and the `domain` column of the `workspaces` table should be `fief.example.com`.

**Reply by MathieuGrosso** [reply]
Hello yes exactly, i have redirect the URL to my domain : fief.our-name.com 
i was unable to make it work like this 

For now i have this when i go the the vm : 

```
~ $ sudo lsof -i -P -n | grep LISTEN
systemd-r  324    systemd-resolve   14u  IPv4  14226      0t0  TCP 127.0.0.53:53 (LISTEN)
systemd-r  324    systemd-resolve   16u  IPv4  14228      0t0  TCP 127.0.0.54:53 (LISTEN)
container  393               root   10u  IPv4  15800      0t0  TCP 127.0.0.1:37331 (LISTEN)
sshd       467               root    3u  IPv4  15553      0t0  TCP *:22 (LISTEN)
sshd       467               root    4u  IPv6  15554      0t0  TCP *:22 (LISTEN)
fief      1197               root   13u  IPv4  17860      0t0  TCP *:80 (LISTEN)
redis-ser 1199               root    6u  IPv4  17820      0t0  TCP *:6379 (LISTEN)
redis-ser 1199               root    7u  IPv6  17821      0t0  TCP *:6379 (LISTEN)
fief      1215               root   16u  IPv4  17840      0t0  TCP *:9191 (LISTEN)


```

But i cant access anywhere the fief server. 


i had put this in the config : 
```
PORT=80
ROOT_DOMAIN= XX.XXX.XX.Xxx
FIEF_DOMAIN= XX.XXX.XX.Xxx
```


i will try with what you said by pointing at the port. 



additional point: 
- if i put after 
fief.example.com on both value, what should i put in port ?

**Reply by frankie567** [reply]
Fief seems to listen on port 80, so looks okay. The port 9191 is a [port opened by the worker to expose Prometheus metrics](https://dramatiq.io/advanced.html#prometheus-metrics), you shouldn't really care about it.

The port doesn't really matter, you can set it on 80 or let it on 8000. It's only important for your reverse proxy, you'll have to tell it that the domain `fief.example.com` should be routed to `your_vm_hostname:8000`.

**Reply by MathieuGrosso** [reply]
i see, its clear like this. 

If i put the domain the indeed i will have to do the redirection via reverse proxy. 

I have first tried to put this :

```
PORT=8000
ROOT_DOMAIN=XX.XXX.XX.Xxx:8000
FIEF_DOMAIN=XX.XXX.XX.Xxx:8000
```

to check if it works like  this but i get a "CSRF a échoué." I think it might be linked to the protocol being http ? Maybe i need to first setup the reverse proxy if i want to do this. 

Thanks a lot :) 

you really helped up getting through this

**Reply by frankie567** [reply]
My pleasure!

Oh yeah, by default all [cookies have the `Secure` flag](https://docs.fief.dev/self-hosting/environment-variables/#csrf-cookie), so it won't work in HTTP unless disabled.

But I **strongly** recommend you to setup SSL anyway, especially if it's already exposed to the internet. Make sure to check this page to make it working properly with Fief: https://docs.fief.dev/self-hosting/deployment/ssl/

**Reply by MathieuGrosso** [reply]
> My pleasure!
> 
> 
> 
> Oh yeah, by default all [cookies have the `Secure` flag](https://docs.fief.dev/self-hosting/environment-variables/#csrf-cookie), so it won't work in HTTP unless disabled.
> 
> 
> 
> But I **strongly** recommend you to setup SSL anyway, especially if it's already exposed to the internet. Make sure to check this page to make it working properly with Fief: https://docs.fief.dev/self-hosting/deployment/ssl/

Yes really clear ! Yes of course I am planning to setup ssl. Issue is as i can only use supervisord & a single dockerfile in compute engine and no docker compose
It might be more tricky to setup traffiek without docker compose. 

Will need to Check

**Reply by frankie567** [reply]
A cloud specialist can probably have more precise advices, but I think their native Load Balancer is the way-to-go on GCP: https://cloud.google.com/load-balancing/docs/https

**Reply by MathieuGrosso** [reply]
Hello, 
i have tried like this but there is an issue : 
--> the http check that needs to run before the ssl redirection is done fails. Indeed, it needs to check on each vm that the route / (or something else, i have tried with login) returns a 200 status. 
YET 
since fief domain was setup not to the IP of each vm but to the redirection when calling one of the route. It returns a 400


```
18:14:49.444 # Redis version=7.0.15, bits=64, commit=00000000, modified=0, pid=8, just started
8:C 16 Feb 2024 18:14:49.444 # Configuration loaded
8:M 16 Feb 2024 18:14:49.449 # Server initialized
2024-02-16 18:14:50,450 INFO success: fief-server entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-02-16 18:14:50,450 INFO success: fief-worker entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-02-16 18:14:50,450 INFO success: redis entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
Main Fief workspace already exists
Main Fief user already exists
Main Fief admin API key not provided in settings. Skipping its creation.
2024-02-16 18:14:59.772 | INFO     | fief.lifespan:lifespan:28 - Fief Server started - {"version": "0.27.0"}
2024-02-16 18:14:59.772 | WARNING  | fief.lifespan:lifespan:31 - Telemetry is enabled.
We will collect data to better understand how Fief is used and improve the project.
You can opt-out by setting the environment variable `TELEMETRY_ENABLED=false`.
Read more about Fief's telemetry here: https://docs.fief.dev/telemetry - {}
2024-02-16 18:15:18.961 | INFO     | uvicorn.protocols.http.httptools_impl:send:496 - 35.191.41.116:53728 - "GET /login HTTP/1.1" 400 - {}
2024-02-16 18:15:19.730 | INFO     | uvicorn.protocols.http.httptools_impl:send:496 - 35.191.41.112:55520 - "GET /login HTTP/1.1" 400 - {}
2024-02-16 18:15:20.193 | INFO     | uvicorn.protocols.http.httptools_impl:send:496 - 35.191.41.114:51194 - "GET /login HTTP/1.1" 400 - {}
2024-02-16 18:15:23.766 | INFO     | uvicorn.protocols.http.httptools_impl:send:496 - 35.191.41.116:36964 - "GET /login HTTP/1.1" 400 - {}
2024-02-16 18:15:24.728 | INFO     | uvicorn.protocols.http.httptools_impl:send:496 - 35.191.41.112:41854 - "GET /login HTTP/1.1" 400 - {}
```

Therefore, the redirection is not done and on my domain fief.XX-XXX.com i get a 502. which is likely because the http check failed before 



https://stackoverflow.com/questions/62744880/error-server-error-the-server-encountered-a-temporary-error-and-could-not-compl



I dont know if your team as a way to avoid this "loophole" or i am wrong somewhere, thanks :)

**Reply by frankie567** [reply]
Hmm, I suggest you to try the path `/openapi.json`, which should respond no matter what once the server is up.

I'm planning to add a proper healthcheck endpoint in the next version ;)

**Reply by MathieuGrosso** [reply]
Ok was able to make it work with the /openai.json ! it made the load balancer work and now it is redirect to the right address at: 
X.XX-XXX.com

I will do some internal docs to explain how to deploy fief on gcp since it was hard. I can send it to you later if you want, you were a really great help !!!


On a side note -  I might still need to setup some variable, since css seems really strange on the address, (X.XX-XXX.com - you will see the same issue if you go to the address i think. I have made the  address reachable from all IP for now as we are testing some thing, but i will  of course change it later ). I dont know if this is related to gcp or to fief directly. 

<img width="743" alt="Capture d’écran 2024-02-17 à 15 46 47" src="https://github.com/fief-dev/fief/assets/76784995/b6c65fe3-cb7f-42e8-b8ea-017f9c07e73a">

**Reply by frankie567** [reply]
Yep, the generated CSS/JS hyperlinks are in HTTP, so the browser refuses to load them if the main website is in HTTPS.

Did you set `FORWARDED_ALLOW_IPS` correctly? 🙂 https://docs.fief.dev/self-hosting/deployment/ssl/

**Reply by MathieuGrosso** [reply]
Yes for sure ! set on *  for testing purpose, but still get a mixed content issue  !

**Reply by frankie567** [reply]
Hmm, then there must be something missing in the GCP Load Balancer. Check if the proxy correctly sets the `X-Forwarded-Proto` header on the request (I think it can be seen in GCP logs).

**Reply by MathieuGrosso** [reply]
I have been digging it, for now i dont see anything about X-forwarded-proto header. 
Possibly we could define the "En-têtes de requêtes personnalisés" or the "En-têtes de réponses personnalisé" to X-Forwaded-Proto:https but i am not sure if this it. 
i will update this when I have found it :) 

Thanks !

**Reply by MathieuGrosso** [reply]
Technically, GCP automatically adds the X-forwarded-proto header to requests so it should work, but it does not appear to be the case

**Reply by louistransfer** [reply]
Hello !
I am also working on this issue with Mathieu and I am trying to access the user page, I believe that the redirection is working but Fief doesn't register the redirection :

curl -v https://fief.xxx/
Trying 31.xx.xxx.xxx:443...
Connected to [fief.xxx](http://fief.xxx) (31.xx.xxx.xxx) port 443 (#0)
ALPN: offers h2,http/1.1
(304) (OUT), TLS handshake, Client hello (1):
CAfile: /etc/ssl/cert.pem
CApath: none
(304) (IN), TLS handshake, Server hello (2):
(304) (IN), TLS handshake, Unknown (8):
(304) (IN), TLS handshake, Certificate (11):
(304) (IN), TLS handshake, CERT verify (15):
(304) (IN), TLS handshake, Finished (20):
(304) (OUT), TLS handshake, Finished (20):
SSL connection using TLSv1.3 / AEAD-CHACHA20-POLY1305-SHA256
ALPN: server accepted h2
Server certificate:
subject: CN=[fief.xxx](http://fief.xxx/)
start date: Feb 16 14:46:43 2024 GMT
expire date: May 16 15:39:37 2024 GMT
subjectAltName: host "[fief.xx](http://fief.xxx/)" matched cert's "[fief.xxx](http://fief.xxx/)"
issuer: C=US; O=Google Trust Services LLC; CN=GTS CA 1D4
SSL certificate verify ok.
using HTTP/2
h2 [:method: GET]
h2 [:scheme: https]
h2 [:authority: [fief.xxx](http://fief.xxx/)]
h2 [:path: /]
h2 [user-agent: curl/8.1.2]
h2 [accept: /]
Using Stream ID: 1 (easy handle 0x15500c600)
GET / HTTP/2
Host: [fief.xxx](https://github.com/orgs/fief-dev/discussions/fief.xxx)
User-Agent: curl/8.1.2
Accept: /
< HTTP/2 307
< date: Mon, 19 Feb 2024 17:32:12 GMT
< server: uvicorn
< location: http://fief.xxx/login
< content-length: 31
< content-type: application/json
< content-security-policy: frame-ancestors 'self'
< x-frame-options: SAMEORIGIN
< referrer-policy: strict-origin-when-cross-origin
< x-content-type-options: nosniff
< permissions-policy: geolocation=(), camera=(), microphone=()
< via: 1.1 google
< x-forwarded-proto: https
< alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
<
Connection #0 to host [fief.xxx](http://fief.xxx/) left intact
{"detail":"Temporary Redirect"}%

**Reply by frankie567** [reply, marked as answer]
@louistransfer Nothing wrong with what you show me here. When accessing the root path `/` without a valid user session, Fief returns a redirection to `/login` (hence the 307 status code). By default, cURL doesn't follow redirections. You can add the `-L` flag to tell it to automatically follow redirections: https://everything.curl.dev/http/redirects#tell-curl-to-follow-redirects

@MathieuGrosso Yes, normally that's something proxy are doing by default, this is quite standard. GCP docs even claim it's the case: https://cloud.google.com/load-balancing/docs/https

...

I just had another look at your Dockerfile. I see that you are copying a `.env` file. Are you setting `FORWARDED_ALLOW_IPS` inside that file? If yes, then it won't work 😅

The fact that Fief is able to parse `.env` file is because we use a specific wrapper for our settings, which is able to fallback to the values in this file **if they are not available in OS environment variables**. However, `FORWARDED_ALLOW_IPS` is not a Fief setting; it's something that's expected by the underlying Python server, Uvicorn. So it *must* be a true OS environment variable (like you set `GCP_PROJECT` at the beginning of your Dockerfile).

**Reply by MathieuGrosso** [reply]
yes it was exactly it, i have set it directly in the supervisord conf to pass it as env variable to the python server and it made it work. I saw it when printing env variable of the process 

Thanks a lto
